### PR TITLE
Sharing data via frontend

### PR DESCRIPTION
**Pull request by acrellin**
This PR introduces a data sharing page where users can select photometry data and spectra to share with other groups/users. The tables are sortable & filterable (with the select all button applying only to filtered results) for easy isolation of specific filters/instruments. A new `/api/sharing` endpoint and associated handler is added, which updates group access to the specified photometry/spectra. Both spectra and photometry can be shared in a single request. Links to the data sharing page are added below the photometry and spectra plots on the source page.

The data model is also updated: spectra now have group-level access control, and `group_ids` is a required field for all `POST` requests to `/api/spectra`.

See screenshot capture below:

share_phot_spec

### CODE REVIEW

**Comment by profjsb** [comment]
Looks good. On the UX it seems there's no way to see what groups that a data point is already shared with--perhaps another column with set of small colored circles representing each group (with a hover-over and a key below the table). We should also think how to _remove_ access to groups and what the permissioning is on this (ie. does someone need to be a group owner of that removed group or simply the original author)?

File: static/js/components/ShareDataForm.jsx [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+              />
+              )}
+            control={control}
+            onChange={([, data]) => data}
+            rules={{ validate: validateGroups }}
+            defaultValue={[]}
+          />
+          <br />
+          <div>
+            <Button
```
Comment (dannygoldstein) [inline static/js/components/ShareDataForm.jsx:201]: Currently there is no error raised if one tries to share a spectrum or photometry point with someone it is already shared with. Is that expected behavior?

File: static/js/components/ShareDataForm.jsx [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+      <div>
+        <form onSubmit={handleSubmit(onSubmit)}>
+          {
+            !!errors.groups &&
+              <FormValidationError message="Please select at least one group/user" />
+          }
+          <Controller
+            name="groups"
+            as={(
+              <Autocomplete
```
Comment (dannygoldstein) [inline static/js/components/ShareDataForm.jsx:177]: Where are the single user groups (in load demo data)?

File: static/js/components/ShareDataForm.jsx [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+  if (!photometry[route.id] || !allGroups || !spectra[route.id]) {
+    return <>Loading...</>;
+  }
+
+  const photRows = photometry[route.id].map((phot) => (
+    createPhotRow(phot.id, phot.mjd, phot.mag, phot.magerr, phot.limiting_mag,
+      phot.instrument_name, phot.filter)));
+
+  const specRows = spectra[route.id].map((spec) => (
```
Comment (dannygoldstein) [inline static/js/components/ShareDataForm.jsx:111]: Just showing the ID and the instrument is not enough to unambiguously identify the spectrum to the user. Perhaps also add the date taken? And an MUI expandable row to show the spectrum plot? Need a way to validate exactly which spectrum is being shared.

File: static/js/components/ShareDataForm.jsx (hunk 1 above)
Comment (acrellin) [reply static/js/components/ShareDataForm.jsx:201]: Yes

File: static/js/components/ShareDataForm.jsx [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    }
+    setIsSubmitting(false);
+  };
+
+  if (!photometry[route.id] || !allGroups || !spectra[route.id]) {
+    return <>Loading...</>;
+  }
+
+  const photRows = photometry[route.id].map((phot) => (
+    createPhotRow(phot.id, phot.mjd, phot.mag, phot.magerr, phot.limiting_mag,
```
[the remark this answers is not in the document]
Comment (acrellin) [reply static/js/components/ShareDataForm.jsx:108]: Will do

File: static/js/components/ShareDataForm.jsx (hunk 2 above)
Comment (acrellin) [reply static/js/components/ShareDataForm.jsx:177]: Looks like they're not being set up properly in data_loader (Stéfan's PR fixes that)

File: static/js/components/ShareDataForm.jsx (hunk 3 above)
Comment (acrellin) [reply static/js/components/ShareDataForm.jsx:111]: Date sounds good -- is a plot really necessary here?

**Comment by dannygoldstein** [comment]
Yes, multiple spectra can be taken on the same date. MUI-datatables has a renderExpandableRow argument that should work fine for this. It could use the Plot component

________________________________

File: static/js/components/ShareDataForm.jsx (hunk 3 above)
Comment (acrellin) [reply static/js/components/ShareDataForm.jsx:111]: @dannygoldstein done. Modified the plot handler to optionally take spec ID so we can render the individual spec plots:


share_spec_plots

File: static/js/components/ShareDataForm.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (acrellin) [reply static/js/components/ShareDataForm.jsx:112]: @dannygoldstein actually, it's gotta be one or the other -- the handler, as you've written it, returns _either_ fluxes or mags, but not both

File: skyportal/handlers/api/sharing.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                schema: Success
+        """
+        data = self.get_json()
+        group_ids = data.get("groupIDs", None)
+        if group_ids is None or group_ids == []:
+            return self.error("Missing required `groupIDs` field.")
+        phot_ids = data.get("photometryIDs", [])
+        spec_ids = data.get("spectrumIDs", [])
+        if not phot_ids and not spec_ids:
+            return self.error("One of either `photometryIDs` or `spectrumIDs` "
```
Comment (stefanv) [inline skyportal/handlers/api/sharing.py:53]: Where are permissions checked to ensure that the user has access to the photometry/spectra/etc. in question?

File: skyportal/models.py [hunk 6]
```diff
@@ -414,6 +418,19 @@ def get_photometry_owned_by_user(obj_id, user_or_token):
 Obj.get_photometry_owned_by_user = get_photometry_owned_by_user
 
 
+def get_spectra_owned_by(obj_id, user_or_token):
+    return (
+        Spectrum.query.filter(Spectrum.obj_id == obj_id)
+        .filter(
+            Spectrum.groups.any(Group.id.in_([g.id for g in user_or_token.groups]))
```
Comment (stefanv) [inline skyportal/models.py:425]: accessible_groups?

File: static/js/components/ShareDataForm.jsx [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+  const onSubmit = async (groupsFormData) => {
+    const selectedPhotIDs = selectedPhotRows.map((idx) => photometry[route.id][idx].id);
+    const selectedSpecIDs = selectedSpecRows.map((idx) => spectra[route.id][idx].id);
+    setIsSubmitting(true);
+    const data = {
+      groupIDs: groupsFormData.groups.map((g) => g.id),
+      photometryIDs: selectedPhotIDs,
+      spectrumIDs: selectedSpecIDs,
+    };
+    const result = await dispatch(sourceActions.shareData(data));
```
Comment (stefanv) [inline static/js/components/ShareDataForm.jsx:101]: LSP warns that this await is not being used correctly; I see we use the same pattern elsewhere.  Let's verify.

File: skyportal/models.py (hunk 6 above)
Comment (acrellin) [reply skyportal/models.py]: Done

File: skyportal/handlers/api/sharing.py (hunk 5 above)
Comment (acrellin) [reply skyportal/handlers/api/sharing.py:53]: Done

**Comment by acrellin** [comment]
Added websocket messages to auto-update phot/spec tables: 
share_phot_update_table

File: static/js/components/ShareDataForm.jsx [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+  { name: "instrument", label: "Instrument" },
+  { name: "filter", label: "Filter" },
+  { name: "groups", label: "Existing Groups" },
+];
+
+const specHeadCells = [
+  { name: "id", label: "ID" },
+  { name: "instrument", label: "Instrument" },
+  { name: "observed", label: "Observed (UTC)" },
+  { name: "groups", label: "Existing Groups" },
```
Comment (dannygoldstein) [inline static/js/components/ShareDataForm.jsx:60]:
```suggestion
  { name: "groups", label: "Currently visible to" },
```

File: static/js/components/ShareDataForm.jsx [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+const photHeadCells = [
+  { name: "id", label: "ID" },
+  { name: "mjd", label: "MJD" },
+  { name: "mag", label: "Mag" },
+  { name: "magerr", label: "Mag Error" },
+  { name: "limiting_mag", label: "Limiting Mag" },
+  { name: "instrument", label: "Instrument" },
+  { name: "filter", label: "Filter" },
+  { name: "groups", label: "Existing Groups" },
```
Comment (dannygoldstein) [inline static/js/components/ShareDataForm.jsx:53]:
```suggestion
  { name: "groups", label: "Currently visible to" },
```

File: static/js/components/ShareDataForm.jsx [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+];
```
Comment (dannygoldstein) [inline static/js/components/ShareDataForm.jsx:54]:
```suggestion
{ name: "owner", label: "Owner" }
];
```

File: static/js/components/ShareDataForm.jsx [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+const specHeadCells = [
+  { name: "id", label: "ID" },
+  { name: "instrument", label: "Instrument" },
+  { name: "observed", label: "Observed (UTC)" },
+  { name: "groups", label: "Existing Groups" },
+];
```
Comment (dannygoldstein) [inline static/js/components/ShareDataForm.jsx:61]:
```suggestion
{ name: "owner", label: "Owner" }
];
```

File: static/js/components/ShareDataForm.jsx (hunk 10 above)
Comment (acrellin) [reply static/js/components/ShareDataForm.jsx:54]: @dannygoldstein we don't store owner/uploader info yet...

File: static/js/components/ShareDataForm.jsx (hunk 9 above)
Comment (acrellin) [reply static/js/components/ShareDataForm.jsx]: Done

File: static/js/components/ShareDataForm.jsx (hunk 8 above)
Comment (acrellin) [reply static/js/components/ShareDataForm.jsx]: Done

**Comment by acrellin** [comment]
@stefanv all of your (and Josh's and Danny's) comments have been addressed. Please merge.

**Comment by profjsb** [comment]
Not clear from the gifs that you have implemented the removal of groups (I see only the addition). What's the thinking on this?

**Pull request merged**